RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+| `/moduleroot`             | Templates get deployed on new module, convert and update |
+| `/moduleroot_init`        | Templates get only deployed when the target module does not yet exist |
+| `/object_templates/*.erb` | Template location for `pdk new ...` generation |
+| `/robocop`                | Custom ruby code to assist in generating rubocop profiles |
+
+**V2 locations**
+
+Where `name` is a unique name for the template
+
+> The `name` should use snake case (lowercase with underscore) to be cross platform compliant. The more descriptive name can be specified in the template.yml file
+
+| Path                               | Description |
+| ---------------------------------- | ----------- |
+| `/<name>/README.md`            (1) | A plain text description of the template. Though not really necessary as most of this would be contained in the `template.yml` file. It is expected that this file will be auto-generated |
+| `/<name>/template.yml`             | YAML file describing the template |
+| `/<name>/template_schema.json` (1) | JSON Schema file describing the valid settings for this template |
+| `/<name>/files/**/*`               | Templates to process for new module, convert and update |
+| `/<name>/files_init/**/*`      (1) | Templates to process when the module does not yet exist |
+| `/<name>/*`                  (1,2) | Additional files used _for_ templating, for example configuration information |
+
+(1) Optional
+
+(2) Ruby files that will be evaluated for the ERB templating process **must** be specified in the `template.yml` file. This is to help (somewhat) safeguard arbitrary Ruby being evaluated. Although it can be argued that ERB templates are also arbitrary code.
+
+> TODO: Does files_init EVEN make sense?!??!
+> Actually it does it in an `always_apply` template
+
+> TODO: What about object_templates?
+
+Note - `/object_templates/*.erb` will still be consumed by the V2 Templating Engine.
+
+> TODO: What about `/rubocop/` ?
+
+### `<name>/template.yml`
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| name (required) | A short descriptive name of the template |
+| description | A long (even multiline) description of the template. Default is the name of the template |
+| type | What type of item this is a template for. Currently only `module` is supported, but in time `control_repo` and others may be added. Default is `module` |
+| evaluate_ruby_files | An array of files to [evaluate](https://ruby-doc.org/core-2.6.4/Kernel.html#method-i-eval) prior to rendering ERB templates. This is used to inject complex logic into the templating system without the troubles of having complex logic in ERB.  For example complex branching based on settings. Default is empty array |
+| default_settings | A YAML hash of the default settings for the template. Default is empty hash |
+| always_apply | Always apply this template. If not specified in the template list, it will be applied last. Default is `false` |
```

Comment:
I can see your point about it being against the opt-in requirement.  I introduced this concept to stop too much boilerplate in the `.sync.yaml`.  For example, the default template set would be:

``` yaml
---
# ...
templates
  - gemfile
  - rubocop
  - git  # This would be for gitignore and gitattributes
  - pdkignore
  - rspec
  - rubocop
  - yard
  - rakefile
  - spec_helper # spec/spec_helper.rb and spec/default_facts.yaml
```

Having to add all of those _every_ time would be a bad experience.  Other possible options:

1. Add this list into a file in the root of the templates directory (e.g. `/templates.yml`) that lists these "Default" templates

2. Use the knockout prefix in the sync.yaml to remove "default" templates  e.g.

``` yaml
---
pdk_templates:
  templates:
    - ---git
```

3. Allow users to ignore `default` templates via a `.sync.yaml` setting e.g.

``` yaml
---
pdk_templates:
  use_default_templates: false
```